### ISSUE TITLE
Implement new loss functions in `train.py`

### ISSUE THREAD

**Issue by jackbaker1001** [XanaduAI/GradDFT#43]
Presently, we have the "energy-only" loss in `functional.py` currently named `default_loss`. We should move this into `train.py` and call it "energy_loss". In total, we should have 3 loss functions:

1. `energy_loss`: can be used with self-consistent and non-self consistent training 
2. `density_loss`: only for use in self-consistent training (and maybe Harris Fouled training if we implement it). Anything to do with density in non-self consistent training doesn't make sense as we never update the 1RDM so the predicted density is static in training.
3. `energy_and_density_loss`: For use in self-consistent training only for same reasons as 2.

There are some more considerations here as to what to do with the `stop_gradient` on the non-XC energy when using the different loss functions, but this will be handled in another issue. 

We also have the "implied self consistency" approach I recommended some time ago. I've attached the whiteboard scrawling for this below:

![PXL_20230823_211058359 MP](https://github.com/XanaduAI/DiffDFT/assets/20515391/225cef7e-ffcb-4f28-9e3c-f868e2e7d015)

Long story short, self consistency can be by passed during training if we calculate the total energy passing it a true many body density. You then match the predicted density one step forward in an SCF cycle to the true density (second term in the equation). This loss has the same minimum point as a full self consistent training.
